sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java

```diff
@@ -211,14 +210,64 @@ public TableType getJdbcTableType()
       final Iterator<Entry<DataSegment, SegmentMetadataHolder>> availableSegmentEntries = availableSegmentMetadata.entrySet()
                                                                                                                   .iterator();
 
+      // in memory map to store segment data from available segments
+      final Map<String, PartialSegmentData> partialSegmentDataMap = new HashMap<>();
+      for (SegmentMetadataHolder holder : availableSegmentMetadata.values()) {
+        final PartialSegmentData data = new PartialSegmentData(
+            holder.isAvailable(),
+            holder.isRealtime(),
+            holder.getNumReplicas(),
+            holder.getNumRows()
+        );
+        partialSegmentDataMap.put(holder.getSegmentId(), data);
+      }
+
       //get published segments from coordinator
       final JsonParserIterator<DataSegment> metadataSegments = getMetadataSegments(
           druidLeaderClient,
           jsonMapper,
           responseHandler
       );
 
-      Set<String> availableSegmentIds = new HashSet<>();
+      final Set<String> segmentsAlreadySeen = new HashSet<>();
+
+      //auth check for published segments
+      final CloseableIterator<DataSegment> authorizedPublishedSegments = getAuthorizedPublishedSegments(
+          metadataSegments,
+          root
+      );
+      final FluentIterable<Object[]> publishedSegments = FluentIterable
+          .from(() -> authorizedPublishedSegments)
+          .transform(val -> {
+            try {
+              if (!segmentsAlreadySeen.contains(val.getIdentifier())) {
+                segmentsAlreadySeen.add(val.getIdentifier());
+              }
+              final PartialSegmentData partialSegmentData = partialSegmentDataMap.get(val.getIdentifier());
+              return new Object[]{
+                  val.getIdentifier(),
+                  val.getDataSource(),
+                  val.getInterval().getStart(),
+                  val.getInterval().getEnd(),
+                  val.getSize(),
+                  val.getVersion(),
+                  val.getShardSpec().getPartitionNum(),
+                  partialSegmentData == null ? 0L : partialSegmentData.getNumReplicas(),
+                  partialSegmentData == null ? -1L : partialSegmentData.getNumRows(),
```

Comment:
I know this was from the original review I wasn't involved with, but is -1 the right value to use here? Can this lead to an incorrect result if you are summing the total number of rows across a bunch of segments and some are returning real values and others are returning -1?

Reply:
That's a good point, so `numRows` can be null, but since longs are not nullable yet, so I was passing `-1L` (my thought was if I keep it 0, it may be misleading that a segment has 0 rows, didn't think about the affect on sum), but now I think, `0L` might make more sense if we treat missing values as zeros.